k9s: v0.29/v0.30 compatibility

LucasWagler opened the pull request:
### Description

- Change file extension for v0.30+ (`.yml` -> `.yaml`)
- Change skin file location for v0.29+ (`XDG_CONFIG_DIR/k9s/skins/`)
- Deprecate `programs.k9s.skin` & replace with `programs.k9s.skins`, allowing multiple skin definitions.

Closes #4855

### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://nix-community.github.io/home-manager/#sec-guidelines

-->

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes. _Individual k9s tests worked; a gpg test is failing when running all tests._

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- ~~If this PR adds a new module~~

  - ~~Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).~~ _I'm happy to be added if it'd be helpful._

#### Maintainer CC

<!--
If you are updating a module, please @ people who are in its `meta.maintainers` list.
If in doubt, check `git blame` for whoever last touched something.
-->

@katexochen @liyangau

LucasWagler commented:
TODO:

- [ ] v0.29 single "default" skin file
- [ ] v0.29 multiple skin files
- [ ] v0.30 file extension changes

LucasWagler commented:
First time contributor. Let me know what I can help clean up & do better!

katexochen commented on modules/programs/k9s.nix:
```diff
@@ -140,31 +159,52 @@ in {
 
-  config = mkIf cfg.enable {
+  config = let
+    isPre0_29 = (builtins.compareVersions cfg.package.version "0.29.0") < 0;
```
Maybe we should rather use [lib.versionAtLeast](https://noogle.dev/f/lib/versionAtLeast).

katexochen commented on modules/programs/k9s.nix:
```diff
@@ -34,7 +34,8 @@ in {
       default = { };
       description = ''
-        Skin written to {file}`$XDG_CONFIG_HOME/k9s/skin.yml`. See
+        Deprecated; see `programs.k9s.skins`.
+        Skin written to {file}`$XDG_CONFIG_HOME/k9s/skins/default.yaml`. See
```
Should we set `settings.k9s.ui.skin = "default"` when unset and the fallback is used? Otherwise the default skin won't be used.

LucasWagler replied:
Definitely. I forgot to implement that.

katexochen reviewed (commented):
Otherwise looks good to me. I had some struggle testing this with v0.30.8 as k9s would just crash all the time. Was able to test using the [unmerged v0.31.5](https://github.com/NixOS/nixpkgs/pull/280280).

@LucasWagler good work, feel free to add yourself to the maintainers if you like!

katexochen commented:
Oh, and please add a NEWS entry to inform about the changes. :)

LucasWagler commented:
In the latest pushes, I reorganized tests into some folders, added a news entry, and added the default skin name logic for 'programs.k9s.settings.k9s.ui.skin`.

It should be ready to ship now!

katexochen reviewed (approved):
LGTM, thanks!

@rycee for last review and merge

LucasWagler commented:
Ran a final formatting pass, as I forgot on the most recent run. I'll set up some pre-commit hooks for myself for the future :+1: 

_Now_ it's ready to go :sweat_smile:

rycee commented on modules/programs/k9s.nix:
```diff
@@ -8,9 +8,9 @@ let
 
 in {
-  meta.maintainers = with maintainers; [ katexochen liyangau ];
+  meta.maintainers = with maintainers; [ katexochen liyangau LucasWagler ];
```
To refer to the HM maintainers file you need to use `lib.hm.maintainers` attribute set:

```suggestion
  meta.maintainers = with maintainers; [ katexochen liyangau hm.maintainers.LucasWagler ];
```

The `lib.maintainers` attribute set refers to the maintainer list of Nixpkgs.

rycee commented on modules/programs/k9s.nix:
```diff
@@ -140,31 +163,67 @@ in {
+      };
+    } // deprecatedSkinOpt // skinFiles;
+
+    warnings = optional (cfg.skin != { }) ''
+      Using programs.k9s.skin is deprecated and will be
+      removed in the future. Please change to using programs.k9s.skins
+      as an attribute set of one or more skin definitions.
+    '';
```
Perhaps it would be sufficient to use, `mkRenamedOptionModule`? Something like

```nix
imports = [
  (mkRenamedOptionModule [ "programs" "k9s" "skin" ] [ "programs" "k9s" "skins" "skin"])
];
```

perhaps? It goes between the `meta.maintainers` and `options.programs.k9s` lines.

LucasWagler replied:
Potentially. The new `skins` option does expect a slightly different structure, though.
Old `skin`: YAML file content
New `skins`: attrset of { fileName = { YAML file content }, fileName2 = { YAML file content } }

LucasWagler replied:
It worked well! Thanks for the elegant solution.

rycee commented on modules/misc/news.nix:
```diff
@@ -1387,6 +1387,15 @@ in
           A new module is available: 'xdg.portal'.
         '';
       }
+
+      {
+        time = "2024-01-17T23:20:41+00:00";
+        message = ''
+          The option `programs.k9s.skin` has been deprecated. It is replaced
+          by `programs.k9s.skins` to allow for multiple skin files to be
+          created.
+        '';
+      }
```
No need for the news entry since the rename warning will be printed anyway.

```suggestion
```

rycee commented on modules/programs/k9s.nix:
```diff
@@ -8,9 +8,13 @@ let
+  ];
 
-  options.programs.k9s = {
+  options.programs.k9s = rec {
```
Shouldn't be necessary to use rec here:

```suggestion
  options.programs.k9s = {
```

rycee commented on modules/programs/k9s.nix:
```diff
@@ -140,31 +163,67 @@ in {
-    xdg.configFile."k9s/plugin.yml" = mkIf (cfg.plugin != { }) {
-      source = yamlFormat.generate "k9s-plugin" cfg.plugin;
+  config = let
+    isAtLeast0_29 = (lib.versionAtLeast cfg.package.version "0.29.0");
+    isAtLeast0_30 = (lib.versionAtLeast cfg.package.version "0.30.0");
```
I don't think you have to worry about these. The package in Nixpkgs unstable is (currently) 0.31.5 and I don't see any earlier version available there so in HM we shouldn't have to worry about earlier versions.

LucasWagler replied:
Perfect, that simplifies things. I did wonder just how backwards-compatible we needed to be in the case of an overridden `programs.k9s.package`.

LucasWagler commented:
@rycee I implemented all of your suggestions. Thanks for taking the time to give a thorough review! Your ideas greatly simplified the changes, and I learned quite a few things to use in the future.

rycee commented:
Thanks! Merged to master now 🙂
